EIPS/eip-4931.md

```diff
@@ -0,0 +1,353 @@
+	function upgrade(address _to, uint256 sourceAmount) external {
+		require(upgradeStatus == true, "SourceUpgrade: upgrade status is not active");
+		(uint256 destinationAmount, uint256 sourceRemainder) = computeUpgrade(sourceAmount);
+		sourceAmount -= sourceRemainder;
+		require(sourceAmount > 0, "SourceUpgrade: disallow conversions of zero value");
+
+		upgradedBalance[msg.sender] += sourceAmount;
+		source.safeTransferFrom(
+			msg.sender,
+			address(this),
+			sourceAmount
+			);
+		destination.safeTransfer(_to, destinationAmount);
+		sourceUpgradedTotal += sourceAmount;
+		emit Upgrade(msg.sender, _to, sourceAmount, destinationAmount);
+	}
+
+	/// @dev A method to conduct a downgrade from destination token to source token.
+	/// The call will fail if downgrade status is not true, if approve has not been called
+	/// on the destination contract, or if destinationAmount is larger than the amount of destination tokens at the msg.sender address.
+	/// If the ratio would cause an amount of tokens to be destroyed by rounding/truncation, the downgrade call will only downgrade
+	/// the nearest whole amount of destination tokens returning the excess to the msg.sender address.
+	///  Emits the Downgrade event
+	/// @param _to The address the source tokens will be sent to upon completion of the downgrade
+	/// @param destinationAmount The amount of destination tokens that will be downgraded
+	function downgrade(address _to, uint256 destinationAmount) external {
+		require(upgradeStatus == true, "SourceUpgrade: upgrade status is not active");
+		(uint256 sourceAmount, uint256 destinationRemainder) = computeDowngrade(destinationAmount);
+		destinationAmount -= destinationRemainder;
+		require(destinationAmount > 0, "SourceUpgrade: disallow conversions of zero value");
+		require(upgradedBalance[msg.sender] >= sourceAmount,
+			"SourceUpgrade: can not downgrade more than previously upgraded"
+			);
+
+		upgradedBalance[msg.sender] -= sourceAmount;
+		destination.safeTransferFrom(
+			msg.sender,
+			address(this),
+			destinationAmount
+			);
+		source.safeTransfer(_to, sourceAmount);
+		sourceUpgradedTotal -= sourceAmount;
+		emit Downgrade(msg.sender, _to, sourceAmount, destinationAmount);
+	}
+}
+```
+
+
+## Security Considerations
+The main security consideration is ensuring the implementation of the interface handles the source tokens during the upgrade in such a way that they are no longer accessible. Without careful handling, the validity of the upgrade may come into question since source tokens could potentially be upgraded multiple times. This is why EIP-XXXX will strictly enforce the use of ```burn``` for source tokens that are burnable. For non-burnable tokens, the accepted method is to send the source tokens to the ```0x00``` address. When the downgrade Optional Ext. is implemented, the constraint will be relaxed, so that the source tokens can be held by the upgrade contract.
```

Comment:
```suggestion
The main security consideration is ensuring the implementation of the interface handles the source tokens during the upgrade in such a way that they are no longer accessible. Without careful handling, the validity of the upgrade may come into question since source tokens could potentially be upgraded multiple times. This is why EIP-XXXX will strictly enforce the use of `burn` for source tokens that are burnable. For non-burnable tokens, the accepted method is to send the source tokens to the `0x00` address. When the downgrade Optional Ext. is implemented, the constraint will be relaxed, so that the source tokens can be held by the upgrade contract.
```

Use single backticks for inline code